File: 1262-greatest-sum-divisible-by-three/1262-greatest-sum-divisible-by-three.java
class Solution {

    public int maxSumDivThree(int[] nums) {
        int[] f = { 0, Integer.MIN_VALUE, Integer.MIN_VALUE };
        for (int num : nums) {
            int[] g = new int[3];
            System.arraycopy(f, 0, g, 0, 3);
            for (int i = 0; i < 3; ++i) {
                g[(i + (num % 3)) % 3] = Math.max(
                    g[(i + (num % 3)) % 3],
                    f[i] + num
                );
            }
            f = g;
        }
        return f[0];
    }
}

File: 1262-greatest-sum-divisible-by-three/1262-greatest-sum-divisible-by-three.cpp
class Solution {
public:
    int maxSumDivThree(vector<int>& nums) {
        vector<int> f = {0, INT_MIN, INT_MIN};
        for (int num : nums) {
            vector<int> g = f;
            for (int i = 0; i < 3; ++i) {
                g[(i + num % 3) % 3] = max(g[(i + num % 3) % 3], f[i] + num);
            }
            f = move(g);
        }
        return f[0];
    }
};